test/test-PreviewProvider.js

```diff
@@ -303,7 +303,58 @@ exports.test_no_metadata_source = function*(assert) {
     response.write(JSON.stringify(fakeResponse));
   });
   yield gPreviewProvider._asyncSaveLinks([fakeSite]);
-  assert.equal(gMetadataStore[0][0].metadata_source, gEmbedlyServiceSource, "metadata source was now computed by Embedly");
+  cachedLink = yield gPreviewProvider._asyncGetEnhancedLinks([fakeSite]);
+
+  assert.equal(gMetadataStore[0][0].metadata_source, gEmbedlyServiceSource, "correct metadata_source in database");
+  assert.equal(cachedLink[0].metadata_source, gEmbedlyServiceSource, "correct metadata_source returned for this link");
+
+  yield new Promise(resolve => {
+    srv.stop(resolve);
+  });
+};
+
+exports.test_prefer_tippytop_favicons = function*(assert) {
+  // we're using youtube here because it's known that the favicon that Embedly
+  // returns is worse than the tippytop favicon - so we want to use the tippytop one
+  const fakeSite = {
+    "url": "http://www.youtube.com/",
+    "title": null,
+    "sanitized_url": "http://www.youtube.com/",
+    "cache_key": "youtube.com/"
+  };
+  const fakeResponse = {
+    "urls": {
+      "http://www.youtube.com/": {
+        "embedlyMetaData": "some embedly metadata",
+        "favicon_url": "https://badicon.com",
+        "background_color": "#BADCOLR"
+      }
+    }
+  };
+
+  // get the tippytop favicon_url and background_color and compare with
+  // what we get from the cached link
+  let tippyTopLink = gPreviewProvider._tippyTopProvider.processSite(fakeSite);
+  let cachedLink = yield gPreviewProvider._asyncGetEnhancedLinks([fakeSite]);
+  let tippyTopFavicon = cachedLink[0].favicon_url;
+  let tippyTopBackgroundColor = cachedLink[0].background_color;
+
+  assert.equal(tippyTopLink.favicon_url, tippyTopFavicon, "TippyTopProvider added a favicon_url");
+  assert.equal(tippyTopLink.background_color, tippyTopBackgroundColor, "TippyTopProvider added a background_color");
+
+  let srv = httpd.startServerAsync(gPort);
+  srv.registerPathHandler(gEmbedlyEndpoint, function handle(request, response) {
+    response.setHeader("Content-Type", "application/json", false);
+    response.write(JSON.stringify(fakeResponse));
+  });
+  // insert a link with some less nice icons in it and get them back
+  yield gPreviewProvider._asyncSaveLinks([fakeSite]);
+  cachedLink = yield gPreviewProvider._asyncGetEnhancedLinks([fakeSite]);
+  tippyTopFavicon = cachedLink[0].favicon_url;
+  tippyTopBackgroundColor = cachedLink[0].background_color;
+
+  assert.equal(tippyTopLink.favicon_url, tippyTopFavicon, "we still took the better tippyTop favicon_url");
+  assert.equal(tippyTopLink.background_color, tippyTopBackgroundColor, "we still took the better tippyTop background_color");
```

Comment:
Can we add an assertion to test that the link object we get back has both the fakeResponse metadata *and* the tippy top data?  If we only check the tippy top fields, we could remove line 351 and this would still pass.